handlers/user.go

```diff
@@ -1298,109 +1336,197 @@ func UserUpdateEmailPost(w http.ResponseWriter, r *http.Request) {
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
-	email := r.FormValue("email")
 
-	if !utils.IsValidEmail(email) {
-		session.AddFlash("Error: Invalid email format!")
+	// check if password is correct
+	formPassword := r.FormValue("password")
+
+	err = bcrypt.CompareHashAndPassword([]byte(userData.Password), []byte(formPassword))
+	if err != nil {
+		session.AddFlash("Error: Invalid credentials!")
 		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
 
-	var existingEmails struct {
-		Count int
-		Email string
+	// validate new email
+	newEmail := r.FormValue("email")
+
+	if userData.Email == newEmail {
+		session.Save(r, w)
+		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
+		return
+	}
+
+	if !utils.IsValidEmail(newEmail) {
+		session.AddFlash("Error: Invalid email format!")
+		session.Save(r, w)
+		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
+		return
 	}
-	db.FrontendWriterDB.Get(&existingEmails, "SELECT email FROM users WHERE email = $1", email)
 
-	if existingEmails.Email == email {
+	emailExists := false
+	err = db.FrontendWriterDB.Get(&emailExists, "SELECT EXISTS (SELECT email FROM users WHERE email = $1)", newEmail)
+	if err != nil {
+		utils.LogError(err, "error checking if email exists", 0, map[string]interface{}{"email": newEmail})
+		session.AddFlash(authInternalServerErrorFlashMsg)
+		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
-	} else if existingEmails.Email != "" {
-		session.AddFlash("Error: Email already exists please choose a unique email")
+	}
+
+	if emailExists {
+		session.AddFlash("Error: Email already exists, please choose a unique email")
 		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
 
-	var rateLimitError *types.RateLimitError
-	err = sendEmailUpdateConfirmation(user.UserID, email)
+	// everything is fine, send confirmation email
+
+	err = sendEmailUpdateConfirmation(user.UserID, newEmail)
 	if err != nil {
-		logger.Errorf("error sending confirmation-email: %v", err)
-		if errors.As(err, &rateLimitError) {
-			session.AddFlash(fmt.Sprintf("Error: The ratelimit for sending emails has been exceeded, please try again in %v.", err.(*types.RateLimitError).TimeLeft.Round(time.Second)))
-		} else {
-			session.AddFlash(authInternalServerErrorFlashMsg)
-		}
+		utils.LogError(err, "error sending email-change confirmation email", 0, map[string]interface{}{"userID": user.UserID, "newEmail": newEmail})
+		session.AddFlash(authInternalServerErrorFlashMsg)
 		session.Save(r, w)
 		http.Redirect(w, r, "/user/settings", http.StatusSeeOther)
 		return
 	}
 
-	session.AddFlash("Verification link sent to your new email " + email)
+	session.AddFlash("Verification link sent to your new email " + newEmail)
```

Comment:
Small ux improvement: tell the user that confirmation link is only valid for 30min

Reply:
fixed, ptal